sycl/include/CL/sycl/detail/program_impl.hpp

```diff
@@ -424,11 +419,13 @@ class program_impl {
   }
 
   bool is_cacheable() const {
-    return is_cacheable_with_build_options(BuildOptions) && AllowKernelsCaching;
+    return is_cacheable_with_options(BuildOptions) &&
+           is_cacheable_with_options(CompileOptions) &&
+           is_cacheable_with_options(LinkOptions) && AllowKernelsCaching;
```

Comment:
Why do we check compile and link options here?

Reply:
It is not allowed to cache kernels/programs when link/compile/build options are not default.
Compile and link options are set only with appropriate call to `compile*` and `link` methods. Once some programmer who does not know about programs/kernels caching feature will set any of these options they will not break anything. So, it is kind of double-check.

Reply:
I think caching for programs built with "compile and link" methods is already disabled by `AllowKernelsCaching`. Am I wrong?

Reply:
BTW, I think `IsKernelsCachingAllowed` more appropriate name for `AllowKernelsCaching` flag.

Reply:
> I think caching for programs built with "compile and link" methods is already disabled by AllowKernelsCaching. Am I wrong?
No, you're not wrong. This is only a double-check.

Reply:
> I think caching for programs built with "compile and link" methods is already disabled by AllowKernelsCaching. Am I wrong?
I see your point. Removed excess checks.

Reply:
You can add an assert to the code to make sure that expectations are met.